src/Sarif/Visitors/RunMergingVisitor.cs

```diff
@@ -87,15 +103,14 @@ public override Result VisitResult(Result node)
                 }
                 else
                 {
-                    ReportingDescriptor rule = node.GetRule(CurrentRun);
                     int newIndex = Rules.Count;
-                    Rules.Add(rule);
+                    Rules.Add(rule ?? node.GetRule(CurrentRun));
 
                     RuleIdToIndex[ruleId] = newIndex;
                     node.RuleIndex = newIndex;
                 }
 
-                node.RuleId = ruleId;
+                node.RuleId = resultRuleId;
```

Comment:
>resultRuleId [](http://example.com/codeflow?start=30&length=12)

why do we have to change this?

Reply:
can't we remove this entire line or just replace if node.ruleid is null/empty?

Reply:
The `resultRuleId` keeps the value from result entity, `ruleId` is the value from referenced rule entity. They may be different if result uses a sub rule id of the rule.
e.g. 
in `rules` array there is a rule id is "SEC105/003", value of `ruleId` should be "SEC105/003"
a result has rule id "SEC105/003/maybenull", which is a sub rule of above rule, `resultRuleId` keeps this id.
`ruleId` should be used as key of the cache `RuleIdToIndex`.
`resultRuleId` used to set on result rule id.

Reply:
updated the implementation pls see latest iteration